### PR TITLE
#155 Adapt Smarty upper/lower functions to be codesafe (e.g. for Turkish locale)

### PR DESCRIPTION
**Pull request by asmecher**
This PR changes the `ucfirst`/`lcfirst`/`strtoupper`/`strtolower` calls in the code that work with method, class, and variable names. These calls are replaced with equivalents (e.g. using `strtr`) that *do not* change depending on the system locale. This resolves a problem in particular with `tr_TR`, for which `strtolower('I') != 'i'`.

### CODE REVIEW

File: libs/sysplugins/smarty_internal_compile_private_special_variable.php [hunk 1]
```diff
@@ -76,7 +77,8 @@ public function compile($args, Smarty_Internal_TemplateCompilerBase $compiler, $
                         $compiler->trigger_template_error("(secure mode) super globals not permitted");
                         break;
                     }
-                    $compiled_ref = '$_' . strtoupper($variable);
+                    $variableToUpper = strtr($variable, 'abcdefghijklmnopqrstuvwxyz', 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
```
Comment (Alexkurd) [inline libs/sysplugins/smarty_internal_compile_private_special_variable.php:80]:
```suggestion
                    $variableUpper = strtr($variable, 'abcdefghijklmnopqrstuvwxyz', 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
```

Comment (asmecher) [reply libs/sysplugins/smarty_internal_compile_private_special_variable.php]: (Thanks for spotting this mistake, @Alexkurd!)

Comment (Alexkurd) [reply libs/sysplugins/smarty_internal_compile_private_special_variable.php]: You're welcome. I've repeated the error earlier with AZ locale.

**Comment by lkppo** [comment]
Hello,

I need also this fix for a Turkish translation of Wikindx users.

Regards,

**Comment by hsarslan** [comment]
I am also waiting for this fix.
Thanks.

**Comment by lkppo** [comment]
[PHP RFC: Locale-independent case conversion](https://wiki.php.net/rfc/strtolower-ascii) corrects it for PHP 8.2.

**Comment by wisskid** [comment]
@lkppo  cool!

**Comment by lkppo** [comment]
Isn't it! I have patched my code with this PR for several months and it is working fine. By the way, I factored with functions ucfirst_ascii (), strtolower_ascii () to minimize the modifications.

**Comment by lkppo** [comment]
Hi @wisskid,

Thanks for merging this PR. My hand patching can go now!

Regards,

**Pull request merged**